lib/src/test/kotlin/app/cash/kfsm/TransitionerTest.kt

```diff
@@ -0,0 +1,225 @@
+    transition.effected shouldBe 0
+    transitioner.postHookExecuted shouldBe 0
+  }
+
+  "returns error on invalid transition" {
+    val transition = transition()
+    val transitioner = transitioner()
+
+    transitioner.transition(Letter(C), transition).shouldBeLeft()
+      .shouldHaveMessage("Value cannot transition [A] to B, because it is currently C")
+
+    transitioner.preHookExecuted shouldBe 0
+    transition.effected shouldBe 0
+    transitioner.postHookExecuted shouldBe 0
+  }
+
+  "returns error when preHook errors" {
+    val error = RuntimeException("preHook error")
+
+    val transition = transition()
+    val transitioner = transitioner(pre = { error.left() })
+
+    transitioner.transition(Letter(A), transition) shouldBeLeft error
+
+    transitioner.preHookExecuted shouldBe 1
+    transition.effected shouldBe 0
+    transitioner.postHookExecuted shouldBe 0
+  }
+
+  "returns error when effect errors" {
+    val error = RuntimeException("effect error")
+
+    val transition = object : Transition<LetterState>(A, B) {
+      override fun effect(value: Value<LetterState>): ErrorOr<Value<LetterState>> = error.left()
+    }
+    val transitioner = transitioner()
+
+    transitioner.transition(Letter(A), transition) shouldBeLeft error
+
+    transitioner.preHookExecuted shouldBe 1
+    transitioner.postHookExecuted shouldBe 0
+  }
+
+  "returns error when postHook errors" {
+    val error = RuntimeException("postHook error")
+
+    val transition = transition()
+    val transitioner = transitioner(post = { error.left() })
+
+    transitioner.transition(Letter(A), transition) shouldBeLeft error
+
+    transition.effected shouldBe 1
+    transitioner.preHookExecuted shouldBe 1
+    transitioner.postHookExecuted shouldBe 1
+  }
+
+  "returns error when preHook throws" {
+    val error = RuntimeException("preHook error")
+
+    val transition = transition()
+    val transitioner = transitioner(pre = { throw error })
+
+    transitioner.transition(Letter(A), transition) shouldBeLeft error
+
+    transition.effected shouldBe 0
+    transitioner.postHookExecuted shouldBe 0
+  }
+
+  "returns error when effect throws" {
+    val error = RuntimeException("effect error")
+
+    val transition = object : Transition<LetterState>(A, B) {
+      override fun effect(value: Value<LetterState>): ErrorOr<Value<LetterState>> = throw error
+    }
+    val transitioner = transitioner()
+
+    transitioner.transition(Letter(A), transition) shouldBeLeft error
+
+    transitioner.preHookExecuted shouldBe 1
+    transitioner.postHookExecuted shouldBe 0
+  }
+
+  "returns error when postHook throws" {
+    val error = RuntimeException("postHook error")
+
+    val transition = transition()
+    val transitioner = transitioner(post = { throw error })
+
+    transitioner.transition(Letter(A), transition) shouldBeLeft error
+
+    transition.effected shouldBe 1
+    transitioner.preHookExecuted shouldBe 1
+  }
+
+  "returns error when persist fails" {
```

Comment:
Should we have a test for when persist throws?

Reply:
added